### ISSUE TITLE
[HOLD for payment 2023-04-04] [$1000] [Tracking issue] Form validation inconsistencies

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#13623]
Tracking issue for consistency issues among platforms for Form validation:

- [ ] https://github.com/Expensify/App/issues/13211
- [x] https://github.com/Expensify/App/issues/12678
- [x] https://github.com/Expensify/App/issues/13024
- [ ] https://github.com/Expensify/App/issues/13847
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010474a084b46fb082</li>
        <li>Upwork Job ID: 1631724449394253824</li>
        <li>Last Price Increase: 2023-03-17</li>
    </ul>
</details>

**Comment by muttmuure** [comment]
OK I'm the bugzero muchacho

**Comment by JmillsExpensify** [comment]
Added the weekly label for now. I'm don't consider this a daily.

**Comment by trjExpensify** [comment]
👋 What's the plan here @luacmartins, and who's taking that on?

**Comment by luacmartins** [comment]
@trjExpensify not sure that we discussed that. Should this be on hold for WAQ or are we moving forward with other follow-up initiatives already?

**Comment by trjExpensify** [comment]
Yeah, but we're talking a matter of days at this point.

**Comment by luacmartins** [comment]
Ok cool. Here are my thoughts/summary on the reported issues so far:

- https://github.com/Expensify/App/issues/12678 - this issue is different from the others and @puneetlath seems to be handling this in [this draft PR](https://github.com/Expensify/App/pull/13994)

- All other issues have the same root cause, the input is auto-focused and then blurred when we click a link/button, which triggers the validation error. I think we need to first decide if form inputs should be auto-focused to begin with. We didn't discuss this in the Form design doc and there are inconsistencies here, e.g. the bank account manual step page has input auto-focus, but none of the other VBA steps do.

**Comment by luacmartins** [comment]
I revived this [Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1672877012494849?thread_ts=1659440463.665489&cid=C01GTK53T8Q).

**Comment by JmillsExpensify** [comment]
Removed the `bug` label and added the `planning` label since this is technically outside of WAQ.

**Comment by luacmartins** [comment]
@trjExpensify we decided that auto-focus will still be a thing for inputs without value, so this issue will persist even after moving to push-to-page, so we should address it now. I think we have a few options, none of which are ideal:

1. Do nothing
2. Remove onBlur validation
3. Remove auto-focus from all inputs
4. Add a delay to the onBlur validation

Option 4 seems to be the only one that allows us to keep the onBlur validation feature and auto-focus behavior, but it's kinda of a hack (although we do have similar instances of it in App). So I'd recommend we go with that.

**Comment by trjExpensify** [comment]
I agree 4 is the best of those. In [this example](https://codepen.io/mudassir0909/pen/qBjvzL) though, is a delay being added or is there another way?

**Comment by luacmartins** [comment]
That is a different way of doing it, but the mousedown/mouseup events are not cross platform so I don't think they'd be an option here.

**Comment by puneetlath** [comment]
Another option might be to have a `shouldValidateOnBlur` prop on the Form component with starts of set to false, but gets set to true the first time the user updates any form value.

**Comment by puneetlath** [comment]
Should we open this up to contributors to present ideas on?

**Comment by trjExpensify** [comment]
This might be a good one for Callstack or someone actually?

**Comment by JmillsExpensify** [comment]
Yeah, I like that idea. This one is more like a project.

**Comment by fabioh8010** [comment]
Hi, I'm Fábio from Callstack and I would like to work on this issue.

**Comment by JmillsExpensify** [comment]
Thanks for taking this on @fabioh8010!

**Comment by fabioh8010** [comment]
**Update:** Sorry for the late, only had time to deep look into this today, I expect to send a proposal tomorrow!

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by luacmartins** [comment]
@fabioh8010 would you mind providing more details on the logic used to mark an input as "dirty"?

**Comment by fabioh8010** [comment]
Hello @luacmartins!

It's going to be very similar to the "touched" logic, as it all takes place in `Form` component. First, We'll declare a `dirtyInputs` internal variable to store the dirty inputs just like the `touchedInputs` we already have.

Then, in `validate()` method we'll filter the validation errors to show only for the dirty inputs instead of the touched ones.

In `onBlur()` callback of each input, we'll remove the `setTouchedInput()` call as we are removing this logic.

In `onInputChange()` of each input, we'll mark that input as dirty before validating it.

Finally, in 'submit()' method all the inputs will be marked as dirty instead of touched to be able to validate the entire form.

I can also share the diff here, but is going to be the PR basically because I only changed the `Form` component.

**Comment by luacmartins** [comment]
@fabioh8010 it seems like we'd loose the initial onBlur validation feature for each input, is that correct?

1. User focuses on input A
2. User focuses on input B
3. Validation for input A is triggered and displays an error saying that input A is required

https://user-images.githubusercontent.com/22219519/220456514-4b5027b7-8190-454c-b523-3a040040727f.mov

**Comment by fabioh8010** [comment]
@luacmartins Yes, we would replace this onBlur validation with real-time validation right at the first focus. If this is not desirable I will try to look alternatives to keep the onBlur validation besides the ones that you guys already suggested in this issue.

**Comment by luacmartins** [comment]
@fabioh8010 we are trying to avoid that. I wonder if we could work out some logic based on the number of touched inputs and if we have values for those inputs or not, something like:
```
if (_.size(touchedInputs) <= 1 && _.size(state.inputValues) === 0) {
    skip validation
}
```

We might have to tweak the logic a bit so we don't skip the first input validation if the user immediately focuses another input.

**Comment by fabioh8010** [comment]
@luacmartins For the checkbox problem I tested the timeout solution and it works pretty well and don't hurt the experience in my opinion.

```
onBlur: () => {
    setTimeout(() => {
        this.setTouchedInput(inputID);
        this.validate(this.state.inputValues);
    }, 100);
},
```

https://user-images.githubusercontent.com/20051562/220989638-7eac4827-9ab9-44d3-9bc8-a9dd6cb366f8.mov

There are other solutions that were discussed in the past e.g using `onMouseDown` and `preventDefault()` in the checkbox, but the timeout approach is definitely the least intrusive one.

But the navigation problem is really tricky, I spent the day trying to find a way to prevent the input validation on blur when the user is navigating or closing the page but I couldn't find a proper solution yet. This issue can happen with any input that wasn't touched, even if I just lose focus of the screen.

I was looking if we have a way of checking if the user is leaving the page or navigating away, although I don't think if it is going to help since the `onBlur` event will certainly fire before we have such indication.

**Comment by luacmartins** [comment]
Thanks for looking into this @fabioh8010! My concern with the timeout solution is that it seems like an antipattern and it wouldn't solve the navigation issue. Is there a way to tell the different between user focus vs auto-focus?

**Comment by fabioh8010** [comment]
We could check if the child input has the `autoFocus` prop enabled and it's not in the `touchedInputs` map yet, something like this:

```
onBlur: () => {
    if (child.props.autoFocus && !this.touchedInputs[inputID]) {
        // first time the auto-focus field is being blurred
        // do something
    }
},
```

**Comment by luacmartins** [comment]
Maybe something like this should work then?

```
onBlur: () => {
    if (child.props.autoFocus && !this.touchedInputs[inputID] && _.isUndefined(this.state.inputValues[inputID])) {
        // input was only auto-focused, we return early to avoid validating it
        return;
    }
    // original logic
},
```

**Comment by fabioh8010** [comment]
This type of logic only works for the auto-focused input, if I focus another one and then focus a checkbox or navigate away we are going to have the same problems.

https://user-images.githubusercontent.com/20051562/221052432-d4c38d4e-4d52-4fb2-96cb-a50f30c13b13.mov

I have been trying to work on a solution for every input, but didn't find one yet besides removing the onBlur validation.

**Comment by luacmartins** [comment]
Good point! IIRC this didn't seem to be an issue on mobile because we don't blur the input on navigation. Is there a way to replicate that behavior on web?

**Comment by fabioh8010** [comment]
The navigation issue also happens in mobile if we dismiss the modal by clicking the X button (same as web).

https://user-images.githubusercontent.com/20051562/221299723-34378f86-aa0a-4dfe-9a37-bf94261cfad0.MOV

The checkbox issue doesn't happen because of how the platform handles this events differently from web.

I'll discuss this issues with our team on Monday to see if we can get more ideas to solve it. Thanks for helping me @luacmartins !

**Comment by luacmartins** [comment]
Good point! Thanks for investigating this and I'm curious to see what other ideas you come up with!

**Comment by tjaniczek** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

User cannot accept terms of service because error message pushes the checkbox away from user's pointer.

### What is the root cause of that problem?

Error message pops up immediately on `onBlur` causing unexpected & immediate layout shifts.

### What changes do you think we should make in order to solve the problem?

We could use animation to ease in the error messages. This would allow enough time to register click on checkbox. It also seems a better user experience to see why and how position of inputs change (if we spread it over time). 

This will also partially fix the issue with feedback being briefly visible when user leaves the page, as the transition should finish before the animation shows visible cue of the error.


https://user-images.githubusercontent.com/7429558/221858435-35bdcb14-c563-46db-b51e-3398e048336d.mov


### What alternative solutions did you explore? (Optional)
 
In my opinion this is a UI / UX type of issue, not a technical one. Therefore we should not look for technical solutions (dirty checks, onBlur hacks) but rather look into the core issue. And the core issue here is that the layout shifts on error.

I briefly explored the idea to keep layout static when the error appears:

https://user-images.githubusercontent.com/7429558/221859882-72a7c059-4cfd-4469-950e-1425f45f17a7.mov


This would require:

- input from the design team (reworking spacing between form elements to fit error labels),
- a considerable amount of dev work to adjust margin/padding for each input type,
- making sure the layout stays consistent even when error messages span multiple lines (i18n)
 
So, even if I see it as superior user experience - I think animations are more pragmatic solution to this issue.

**Comment by luacmartins** [comment]
Thanks for the proposal @tjaniczek! I think your approach is similar to the delay (timeout) approach, with a different UI. However, I think that both of those approaches are more workarounds than a solution to the issue.

**Comment by luacmartins** [comment]
I think at this point we haven't discarded any options -- they just don't seem like the "right" solution yet. We'll definitely re-evaluate if we can't come up with a better solution.

**Comment by tjaniczek** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

There are inconsistencies between platforms on when forms are validated. On mobile user is able to select checkbox without loosing focus on auto-selected text field. On web, each click outside of text fields causes `onBlur` event and triggers validation.

### What is the root cause of that problem?

Forms behave differently between mobile & web due to `keyboardShouldPersistTaps` behaviour not existing in web platform (https://github.com/necolas/react-native-web/issues/1673#issuecomment-658968448)

### What changes do you think we should make in order to solve the problem?

We could use `event.relatedTarget` property to iron out the differences between platforms. The best approximation I found so far is to skip validation when:
- the field was auto-focused
- user left the field empty
- user clicked anywhere but other form field (https://github.com/Expensify/App/issues/13623#issuecomment-1439080872)

Please see attached video to check the results:

https://user-images.githubusercontent.com/7429558/222377985-b2d3bc1e-bc31-491d-a919-3195cd26f653.mov

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~010474a084b46fb082

**Comment by MelvinBot** [comment]
Current assignee @muttmuure is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Current assignee @luacmartins is eligible for the External assigner, not assigning anyone new.

**Comment by luacmartins** [comment]
Thanks for the proposal @tjaniczek, that's an interesting solution. I think that we should still validate when the user clicks on other form fields whether they left the field empty or not. I also see the checkbox issue is still there.

**Comment by luacmartins** [comment]
Made this issue external so we can get more proposals

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We want to make a consistent behavior between native and web related to how we run the validation. 
On web, the validation runs when we press the checkbox, but the checkbox does not get clicked. The validation also runs when we press back, close, or help.
On native, the only issue I see is the validation runs when we press the close or help button because the back button does not animate when navigating back, so we can't really see it. It is very easy to reproduce with the help button because we can go back later to see whether the validation runs or not.

### What is the root cause of that problem?
On native, when we are navigating away, the screen will get blurred, so does the input.
On web, the input will get blurred every mouse down on every other element.

### What changes do you think we should make in order to solve the problem?


I don't think there is a single solution that is cross-platform, so I will handle it for each platform.

Native
We already know that the text input is blurred because we are navigating away. So, we can simply check inside the `onBlur`, is the screen is currently focused? If not, just return early. To get the focus state, we can use the `withNavigationFocus` HOC.

Web
We can make use of the `event.relatedTarget` property from the `onBlur` callback. Related target will tell which element that cause the blur (or the next element that gets the focus) and we can check if the element that cause the blur is a focusable element. However, if it's not a focusable element by default, the related target will return null, with an exception for an element that has `tabindex` attribute set.

https://stackoverflow.com/a/42764495/11488790

In our app, the button and checkbox are converted into a div with a `tabindex` set to either 0 or -1, while the text input still uses `input`. Because `input` is also a focusable element, we should exclude it from the related target, along with `select` and `textarea`.

So, combining for both platforms, the condition will look like this:

```
if (isNotFocused or (relatedTarget exists and relatedTarget.tagName is not one of input, select, textarea)) then return early
```

_But this is not perfect for web because even though the validation does not run when we press the checkbox, the text input will still lose its focus, but I think this is okay_

**Comment by luacmartins** [comment]
Thanks for the proposal @bernhardoj! @rushatgabhane I'd be curious for your thoughts on the several proposals we received on this issue so far.

**Comment by luacmartins** [comment]
No worries! What do you think of the proposals so far?

**Comment by luacmartins** [comment]
No updates yet because of ECX

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by luacmartins** [comment]
Summarizing the progress so far:

Issue: Validations are being triggered and showing errors during back navigations or focusing other inputs because the inputs are being validated on blur.

Solutions:

1. Do nothing and live with the bugs above
2. Remove onBlur validation (empty inputs wouldn't be validated until the user submits the form)
3. Remove auto-focus from all inputs (wouldn't solve for when users select an input but don't enter any values)
4. Add a delay to the onBlur validation (anti-pattern using timeout)
5. Add an animation for the onBlur validation (similar to the solution above, but without using setTimeout)
6. Add [shouldValidateOnBlur](https://github.com/Expensify/App/issues/13623#issuecomment-1423287126) prop to validate onBlur only after user enter a value (only solves for when forms are empty)
7. Prevent onBlur validation for auto focused inputs (doesn't solve for non-auto focused inputs)
8. Use `event.relatedTarget` as described [here](https://github.com/Expensify/App/issues/13623#issuecomment-1454559225)

**Comment by luacmartins** [comment]
Out of all options, I think 8 might be the best one. @rushatgabhane what do you think?

**Comment by luacmartins** [comment]
Thanks for the input @rushatgabhane! I don't think we'll be able to add `keyboardShouldPersistTaps` since to your point that behavior doesn't exist on web.

I do think that the checkbox and link issues is bad UX so we'd need to fix that. Out of all solutions [listed above](https://github.com/Expensify/App/issues/13623#issuecomment-1476455411), I think that the timeout one seems the simplest one so I'd recommend we go with that.

**Comment by luacmartins** [comment]
Cool! @fabioh8010 wanna give [that solution](https://github.com/Expensify/App/issues/13623#issuecomment-1478343322) a try?

**Comment by fabioh8010** [comment]
@luacmartins Prepared a PR for that: https://github.com/Expensify/App/pull/16444

**Comment by MelvinBot** [comment]
@fabioh8010, @luacmartins, @rushatgabhane, @muttmuure Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by muttmuure** [comment]
I don't think I'm required to pay here, as the payment is being handled in the individual issues

**Comment by muttmuure** [comment]
Deleted my earlier comment as there is a linked PR (sorry for my misunderstanding).

I'm going to clarify with Carlos if we have all of the solutions we want, and then use this Upwork job to pay: https://www.upwork.com/jobs/~010474a084b46fb082

**Comment by luacmartins** [comment]
I think we are done. Closing this issue, please reopen if we missed anything!
